### PR TITLE
feat(appset): implement ResourceTree for ApplicationSets

### LINKED ISSUE #15697: Add ResourceTree functionality for AppSet
# Summary

A ResourceTree on the AppSet is needed, similarly to the one existing for Applications, to enable AppSet support in the UI. 
The AppSet resource tree is basically the apps that have an ownerRef of the AppSet, similar to the resource tree returned for app-of-apps. 

# Motivation

To be able to display the AppSet natively in the UI, treating it like an app-of-apps with regard to it's resource tree. 

# Proposal

This is the implementation for app resource tree:
https://github.com/argoproj/argo-cd/blob/3557ec57dc87b2570cf9beaaf8b07bb419883a45/server/application/application.go#L1379

The only resources in the tree for AppSet should be apps.
The implementation should go here :

https://github.com/argoproj/argo-cd/blob/master/server/applicationset/applicationset.go

Here is where the API ResourceTree endpoint for AppSet should probably be:

https://github.com/argoproj/argo-cd/blob/3557ec57dc87b2570cf9beaaf8b07bb419883a45/server/applicationset/applicationset.proto

For possibly future implementation (leaving it here so we don't forget, if it would indeed be needed, will open a separate issue for that):
_After further discussion with @crenshaw-dev 
The CRD of AppSet needs to change to contain the Resources field in the status, similarly to Application. 
The AppSet controller should populate this new field, with the apps that the AppSet generates, in a format similar to the one that the Application controller is populating._

### PR DESCRIPTION
**Pull request by alexymantha**
Part of the effort to add support for ApplicationSets in the UI, adds a ResourceTree endpoint similar that will be used to display the tree of an ApplicationSet and its children Applications. 

As discussed in the linked issue, this uses the ApplicationSet status to track the children and the tree is built based on the data contained in the status.

The progressive sync feature added an ApplicationSetApplicationStatus field that is used to track the status and progress of the sync of the children's applications. I tried to reuse what the progressive sync was already doing. See comments for more details on the refactor.

Closes #15697 

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them. (UI will be implemented in another PR)
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

### CODE REVIEW

**Comment by reggie-k** [comment]
Thanks for being so quick with this!
Can you elaborate about the progressive sync changes?

**Comment by alexymantha** [comment]
> Can you elaborate about the progressive sync changes?

Yes! I tried to reuse the ApplicationSetApplicationStatus that was previously only used for the progressive sync so I had to make some changes to the progressive sync:
- moved some of the logic that was handling the status update outside of the feature flag.
- moved the progressivesync-specific fields to a subfield of the ApplicationSetApplicationStatus.
- change to use a status map instead of a slice to make it easier to update the status along the way since it's not only being updated by the progressive sync anymore.

Most of these changes could be avoided if I didn't try to reuse ApplicationSetApplicationStatus but I think it makes sense to make the status more generic and not depend on the progressive sync.

**Comment by reggie-k** [comment]
Thanks.
Let's wait for Michael's review.
And I agree that only nodes are relevant for AppSet, orphan nodes are not.

**Comment by reggie-k** [comment]
@alexymantha is this waiting for a review or still in progress?

File: pkg/apis/application/v1alpha1/applicationset_types.go
Comment (crenshaw-dev) [inline pkg/apis/application/v1alpha1/applicationset_types.go:795]: Would it be possible to flatten this back out? Nesting the progressive sync fields introduces a world of backwards-compatibility pain.

Comment (alexymantha) [reply pkg/apis/application/v1alpha1/applicationset_types.go:795]: Sure! However, is there another way to clarify what fields are specifically for the progressive sync? I fear the `Status` and `Message` that are specifically for the progressive sync will get confused with the App's `Status` and `Message` which are under `Health`

Comment (crenshaw-dev) [reply pkg/apis/application/v1alpha1/applicationset_types.go:795]: You could prefix the field names with `ProgressiveSync` but leave the `json` tags as-is. That would clarify the Go code without breaking the k8s or Argo API contracts.

File: pkg/apis/application/v1alpha1/applicationset_types.go
Comment (crenshaw-dev) [inline pkg/apis/application/v1alpha1/applicationset_types.go:801]: To avoid a breaking gRPC change, keep the number in the `protobuf` tag for each of these the same and add the new fields after these. It might read a bit wonky, but worth it to avoid the breaking API change.

Comment (alexymantha) [reply pkg/apis/application/v1alpha1/applicationset_types.go:801]: Ah ok thanks for the info! I was wondering if it had an impact

File: pkg/apis/application/v1alpha1/applicationset_types.go
Comment (alexymantha) [reply pkg/apis/application/v1alpha1/applicationset_types.go]: Done!

File: pkg/apis/application/v1alpha1/applicationset_types.go
Comment (alexymantha) [reply pkg/apis/application/v1alpha1/applicationset_types.go]: Done!

File: applicationset/controllers/applicationset_controller_test.go
Comment (alexymantha) [inline applicationset/controllers/applicationset_controller_test.go:4749]: Responsibility of removing the status moved out of the progressive sync

File: applicationset/controllers/applicationset_controller_test.go
Comment (alexymantha) [inline applicationset/controllers/applicationset_controller_test.go:5842]: New tests start here, the rest is just the renaming of the progressive sync fields and change to status map

**Comment by alexymantha** [comment]
I think the PR is ready for feedback. I'm also working on adding an E2E to test the status updates

**Comment by reggie-k** [comment]
@alexymantha can you update this branch from upstream master?

**Comment by alexymantha** [comment]
@reggie-k It should be good to go!

**Comment by reggie-k** [comment]
Thanks a lot! I will test the changes.
As per the discussion with @crenshaw-dev, I will merge these changes into
my original PR, also the other PRs you are working on, so that everything
gets reviewed at once.

Can you also add changes in the CLI for the AppSet resourc-tree?

**Comment by reggie-k** [comment]
As per discussion with @crenshaw-dev, relevant CLI changes can be added later on separately.

**Comment by reggie-k** [comment]
Verified - the missing fields are now present, working as expected.
Thanks @alexymantha!

**Comment by csantanapr** [comment]
what are the next steps to merge this @alexymantha and @reggie-k, just curious on this because there is another PR I'm interested https://github.com/argoproj/argo-cd/pull/16781

**Comment by alexymantha** [comment]
IIUC @crenshaw-dev suggested to merge it alongisde https://github.com/argoproj/argo-cd/pull/15731 to make it easier for the reviews but this PR is standalone and could be merged now if needed. It's up to the maintainers

**Review by todaywasawesome — APPROVED** [review]
@crenshaw-dev I recommend moving ahead with this PR standalone unless you feel differently. 

Sorry @alexymantha it needs merge conflicts fixed again.

**Comment by alexymantha** [comment]
> Sorry @alexymantha it needs merge conflicts fixed again.

Done!

**Review by agaudreault — APPROVED** [review]

**Comment by Tyrael** [comment]
hi, what is the status here? any reason why this isn't merged yet?

**Review by crenshaw-dev — APPROVED** [review]
Code looks good, local test checks out. Thanks, @alexymantha!

**Comment by crenshaw-dev** [comment]
Would you mind rebasing and re-running codegen?

**Comment by alexymantha** [comment]
Thanks @crenshaw-dev! Everything is up-to-date now

**Comment by Tyrael** [comment]
Thanks!

**Comment by reggie-k** [comment]
Thank you @alexymantha and @crenshaw-dev!

**Pull request merged**
